### PR TITLE
Split `init_module` into `init` + `sharded_model`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Adds a single context manager (`fabric.init()`) that supports instantiation on device and with a specific dtype

```python
import torch
from lightning import Fabric

fabric = Fabric(accelerator="cuda", precision="64-true")
with fabric.init():
    x = torch.zeros(1)
    print(x.device)  # cuda
    print(x.dtype)  # 64
```

This can be useful for cases like non-modules, or to circumvent issues that arise from the `sharded_model` context manager

Resolves https://github.com/Lightning-AI/lightning/pull/17462#discussion_r1177031178

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/fabric/strategies/strategy.py [hunk 1]
```diff
@@ -383,7 +394,7 @@ class _Sharded(ABC):
 
     @abstractmethod
     @contextmanager
-    def module_sharded_context(self) -> Generator:
```
Comment (carmocca) [inline src/lightning/fabric/strategies/strategy.py:386]: This is in a proteted interace, so we are free to make this change

**Review by Borda — APPROVED** [review]

**Review by lantiga — APPROVED** [review]
Looks good, I'm only wondering whether we should rename `init_module` to `init_and_shard` at this point?

**Comment by carmocca** [comment]
> I'm only wondering whether we should rename init_module to init_and_shard at this point?

I would keep the current name because whether we shard or not depends on the strategy used and because having `module` clearly signals that this is meant for module initialization

**Comment by awaelchli** [comment]
@carmocca I was under the impression that we would hold off on this since there were some open questions around use cases and APIs. I've outlined my concerns here in this discussion issue #17581 touching on mostly the points discussed in the meeting.

**Comment by carmocca** [comment]
@awaelchli The PR has been waiting for reviews for a week. I'll follow up on #17581. No need to worry, we can iterate further

**Pull request merged**